js/RecentDonations.js

```diff
@@ -0,0 +1,196 @@
+                messageBeingRead = idx
+              }
+
+              break
+            }
+          }
+         
+        }
+      }
+    }
+
+    // Animate a new message onto the page.
+    var displayMessage = function(message) {
+      $('#message').fadeOut(500, function() { 
+        $scope.message = message
+        $scope.$apply()
+        $('#message').fadeIn(500)
+      })
+    }
+
+    // Gets the name (and IDs) of the requested user(s)
+    var getParticipantIDs = function(then) {
+      if (isTeam) { 
+        // We need the name of the team, and the list of members.
+        GLOBALS.GET_TEAM_INFO(uid, function(data) {
+          $scope.displayName = data.name || "(Invalid ID)"
+          
+          // Display the name to the UI.
+          $scope.$apply()
+        }, function (error) {
+          showError("Could not get team info", error)
+        })
+
+        GLOBALS.GET_TEAM_PARTICIPANTS_INFO(uid, function(data) {
+          members = $.map(data, function(x,i) {
+            return x.participantID
+          })
+
+          then()
+        }, function (error) {
+          showError("Could not get team participant info", error)
+        })
+
+      } else {
+
+        // This is an individual
+        members = [uid]
+        
+        GLOBALS.GET_PARTICIPANT_INFO(uid, function(data) {
+          $scope.displayName = data["firstName"] || "(Invalid ID)"
+          dataChanged()
+        }, function(error) {
+          $scope.displayName = "?"
+          showError("Couldn't get participant info", error)
+        })
+
+        then()
+      }
+    }
+
+    // When info about a single participant arrives, add their donations to our list.
+    var handleParticipantDonationInfo = function(data) {
+      // Go through the array of contributors for this user, and add them to our array
+      var newContributions = $.map(data, function(x, i) {
+        return {
+          name: x.donorName,
+          amount: "$" + x.donationAmount.toFixed(2),
+          message: x.message,
+          date: Date.parse(x.createdOn)
+        }
+      })
+
+      contributions = contributions.concat(newContributions)
+
+      // Successful get - decrement our "semaphore"
+      waitingOn = waitingOn - 1;
+      dataChanged()
+    }
+
+    // Call this to refresh data from all our participants.
+    var update = function() {
+      $scope.errors = []
+      contributions = []
+
+      // This is how many callbacks we expect before we can apply.
+      waitingOn = members.length
+
+      for (i = 0; i < members.length; i++) {
+        GLOBALS.GET_PARTICIPANT_DONATION_INFO(members[i], handleParticipantDonationInfo, function(error) {
+          // Failed get - decrement our "semaphore"
+          waitingOn = waitingOn - 1
+          showError("Couldn't get donation info", error)
+        })
+      }
+
+    }
+
+    // Get our IDs and start the update cycle.
+    getParticipantIDs(function() {
+      update()
+      // Get new data from the server every 10s
```

Comment:
...right. Also, now that I'm thinking about it, this should probably be URL-overridable too